mixins/property-required/test/property-required-mixin.test.js

```diff
@@ -0,0 +1,219 @@
+import { createDefaultMessage, createInvalidPropertyTypeMessage, createUndefinedPropertyMessage, PropertyRequiredMixin } from '../property-required-mixin.js';
+import { defineCE, expect, fixture } from '@brightspace-ui/testing';
+import { LitElement } from 'lit';
+
+const tagValid = defineCE(
+	class extends PropertyRequiredMixin(LitElement) {}
+);
+
+const tagString = defineCE(
+	class extends PropertyRequiredMixin(LitElement) {
+		static properties = {
+			attr: { type: String }
+		};
+		constructor() {
+			super();
+			this.addRequiredProperty('attr');
```

Comment:
An alternate API I considered would be to make it part of the property defintion, like this:

```javascript
static properties = {
  attr: { type: String, required: true }
};
```

It's a nice API, but doesn't leave any room for custom validators or messages. We _could_ optionally make it an object:

```javascript
static properties = {
  attr1: { type: String, required: true },
  attr2: {
    type: String,
    required: {
      validator: () => {...},
      message: () => {...}
    }
};
```

Thoughts?

Reply:
I don't have strong feeling one way or the other. I tend to cringe a bit when property definitions get super bloated, but having it all in one place is nice, and it needs to get defined _somewhere_. Mixing with the Lit-specific API feels sorta uncomfortable, but the mixin is Lit-specific anyway, so it's not like a non-Lit component would be able to use this, if such a thing existed in our system. 🤷‍♂️

Reply:
I wonder if Lit would ever introduce `required` as a property of properties. 🤔

Reply:
Yeah. Similarly, the properties could be prefixed with `d2l` but I'm not in love with that either.

My feeling is that is were worried about it, we should just avoid mixing with their API altogether. I'm not particularly worried, but it could happen. Or unlikely as it might be, Lit could one day do something internally to mangle these.

Reply:
Agreed. OK I'll put up a PR those shows what leveraging the existing properties might look like... if it's feasible technically I think I'd prefer that.

Reply:
We're going to go with the alternate approach, which has now been merged into this branch.

Reply:
BTW I know I set a precedent on overloading this `properties` data, but I've been planning to leave it again - if we move the "providers" thing to a controller instead of a mixin, then it'll make sense to declare the "fromProvider" properties when creating the controller (i.e. in the constructor).

Reply:
Cool, yeah I think controllers make a lot of sense for your case since provider properties are by their nature not something consumers of the component should/can be setting anyway.